toolkits/spotify/tests/test_player.py

```diff
@@ -0,0 +1,652 @@
+    mock_response.json.return_value = {
+        "device": {
+            "id": "1234567890",
+            "is_active": True,
+            "name": "Test Device",
+            "type": "Computer",
+        },
+        "currently_playing_type": "track",
+        "is_playing": True,
+        "progress_ms": 10000,
+        "message": "Playback started",
+    }
+    mock_httpx_client.request.return_value = mock_response
+
+    response = await get_playback_state(context=tool_context)
+
+    assert response["device_id"] == "1234567890"
+    assert response["device_name"] == "Test Device"
+    assert response["is_playing"] is True
+    assert response["progress_ms"] == 10000
+    assert response["message"] == "Playback started"
+
+
+@pytest.mark.asyncio
+async def test_get_playback_state_playback_not_active(tool_context, mock_httpx_client):
+    mock_response = MagicMock()
+    mock_response.status_code = 204
+    mock_httpx_client.request.return_value = mock_response
+
+    response = await get_playback_state(context=tool_context)
+
+    assert response["is_playing"] is False
+
+
+@pytest.mark.asyncio
+async def test_get_playback_state_too_many_requests_error(tool_context, mock_httpx_client):
+    mock_response = MagicMock()
+    mock_response.status_code = 429
+    mock_response.raise_for_status.side_effect = httpx.HTTPStatusError(
+        "Too Many Requests", request=MagicMock(), response=MagicMock(status_code=429)
+    )
+    mock_httpx_client.request.return_value = mock_response
+
+    with pytest.raises(ToolExecutionError):
+        await get_playback_state(context=tool_context)
+
+
+@pytest.mark.asyncio
+async def test_get_currently_playing_success(tool_context, mock_httpx_client):
+    mock_response = MagicMock()
+    mock_response.status_code = 200
+    mock_response.json.return_value = {
+        "device": {
+            "id": "1234567890",
+            "is_active": True,
+            "name": "Test Device",
+            "type": "Computer",
+        },
+        "currently_playing_type": "track",
+        "is_playing": True,
+        "progress_ms": 10000,
+        "message": "Playback started",
+    }
+    mock_httpx_client.request.return_value = mock_response
+
+    response = await get_currently_playing(context=tool_context)
+
+    assert response["device_id"] == "1234567890"
+    assert response["device_name"] == "Test Device"
+    assert response["is_playing"] is True
+    assert response["progress_ms"] == 10000
+    assert response["message"] == "Playback started"
+
+
+@pytest.mark.asyncio
+async def test_get_currently_playing_playback_not_active(tool_context, mock_httpx_client):
+    mock_response = MagicMock()
+    mock_response.status_code = 204
+    mock_httpx_client.request.return_value = mock_response
+
+    response = await get_currently_playing(context=tool_context)
+
+    assert response["is_playing"] is False
+
+
+@pytest.mark.asyncio
+async def test_get_currently_playing_too_many_requests_error(tool_context, mock_httpx_client):
```

Comment:
might be worth exploring if we could parameterize some of these tests. For example, parameterize by tool name for all of the tests that return 429.

Reply:
There's some of repetition indeed. Pushed parameterization for a few tests